[B+C] API for Pathfinders. Adds BUKKIT-1710

tomsik68 opened the pull request:
### The Issue

As all of us know, minecraft introduced a new API for mob Pathfinding. It was quite a long time ago and Bukkit is still missing support for Pathfinders. 
### Justification for this PR

We need an API for Pathfinders for a lots of awesome things. 
Like you can set Zombies not to break doors or you can have bunch of mobs following you,  change how monsters attack, change how fast mobs move, make mobs look at you.
### PR Breakdown
- Interface **Controllable**
  - Provides methods for looking at object, jumping and moving using vanilla jumpController, lookController and navigator from net.minecraft.server.Entity  
  - Methods to alter path finding like setAvoidWater, setBreakDoors, setAvoidSun which call vanilla methods from Navigator class.
- Interface **PathfindingEntity**
  - Pathfinder manipulation methods including clearPathfinders, which clears list a and b from goalSelector and targetSelector of EntityLiving and attachPathfinder, which uses PathfinderWrapper in org.bukkit.craftbukkit.ai to add to Entity's task list. Either goalSelector or targetSelector depending on what's specified in Pathfinder class.
  - **This interface is implemented by Animals and chosen types of other mobs**
- Interface **Pathfinder**
  - Pathfinder interface with few functions. This is whole new thing which plugin devs can use to change behavior of Entities.
  - Used by PathfinderWrapper
- Class **PathfinderWrapper**
  - Wrapper for Pathfinders
  - Extends minecraft's internal PathfinderGoal class and its field impl can store implemenetation of Pathfinder interface
  - When minecraft calls methods from it, methods from impl will be called
### Relevant PRs
- https://github.com/Bukkit/CraftBukkit/pull/1128 - My PR with the API created
### Testing material
- PathfinderTest - If you right-click a living entity it will start following you  
  - Plugin JAR - https://dl.dropbox.com/u/29628587/PathfinderTest.jar
  - Source - https://github.com/tomsik68/PathfinderTest
### JIRA Ticket
- https://bukkit.atlassian.net/browse/BUKKIT-1710 - I'm not creator of the ticket, but there's a link to this PR in comments

tomsik68 committed:
Added pathfinders API. Adds BUKKIT-1710
This API can be used by plugins to change behavior of Living Entities in
certain situations. Pathfinders will however only be active until the
server stops. Plugins need to take control of this.

andrepl commented on src/main/java/org/bukkit/ai/Pathfinder.java:
```diff
@@ -0,0 +1,50 @@
+     * @return Entity controlled by the pathfinder
+     */
+    public LivingEntity getEntity();
+    /** Sets entity to be controlled by the pathfinder
```
all of the javadocs are formatted improperly. should be:
    /**
     \* Description Goes Here
     */

you've got the description on the `/**` line.

tomsik68 replied:
Is it better now? :)

andrepl commented on src/main/java/org/bukkit/ai/Pathfinder.java:
```diff
@@ -0,0 +1,53 @@
+     * Sets entity to be controlled by the pathfinder
+     * @param entity
+     */
+    public void setEntity(LivingEntity entity);
```
There should be a blank line between every method. as well as a blank line between the javadoc's first line and any others. eg:

```
/**
 * Sets entity to be controlled by the pathfinder
 *
 * @param entity
 */
```

the space before "Sets" is important, as is the blank line between, in fact the dev team might want a <p> there instead, someone else will know.

natemort commented:
Your javadocs could use some improvement.  Additionally, we no longer use @author annotations.

andrepl commented on src/main/java/org/bukkit/ai/Pathfinder.java:
```diff
@@ -0,0 +1,53 @@
+     * @return If pathfinder should continue execution(if player is in required range etc.)
+     */
+    public boolean canContinue();
+    /** Called when pathfinder execution stops
```
Missed one here

natemort replied:
When does pathfinder execution stop?

andrepl commented on src/main/java/org/bukkit/entity/Intelligent.java:
```diff
@@ -0,0 +1,95 @@
+     */
+    public boolean canBreakDoors();
+    /** 
+     *Sets if this entity can break doors
```
- Sets
  ^ needs space

andrepl commented on src/main/java/org/bukkit/entity/Intelligent.java:
```diff
@@ -0,0 +1,95 @@
+     */
+    public void attachPathfinder(Pathfinder pf);
+    /**
+     *Deletes all the pathfinders(even the vanilla ones)
```
Space here too. \* Deletes

natemort commented on src/main/java/org/bukkit/ai/Pathfinder.java:
```diff
@@ -0,0 +1,53 @@
+    public boolean canStart();
+    /**
+     * 
+     * @return If pathfinder should continue execution(if player is in required range etc.)
```
How frequently is this checked?  When is this checked?

andrepl commented on src/main/java/org/bukkit/entity/Intelligent.java:
```diff
@@ -0,0 +1,95 @@
+     * @param speedPitch
+     * @param speedYaw
+     */
+    public void lookAt(double x,double y,double z,float speedPitch,float speedYaw);
```
spaces after commas, everywhere.

natemort commented on src/main/java/org/bukkit/ai/Pathfinder.java:
```diff
@@ -0,0 +1,53 @@
+     */
+    public int getPriority();
+    /**
+     *Performs a logical update of this pathfinder.
```
How frequently is this called?  What should a pathfinder do on an update?

natemort commented on src/main/java/org/bukkit/ai/Pathfinder.java:
```diff
@@ -0,0 +1,53 @@
+     */
+    public PathfinderType getType();
+    /**
+     *Gets priority of this pathfinder. 0 = highest
```
What does the priority determine?

natemort commented on src/main/java/org/bukkit/ai/PathfinderType.java:
```diff
@@ -0,0 +1,19 @@
+    Goal,
+    /**
+     *Targetting pathfinders help monsters & golems to choose targets
+     *  Note: You should call {@link EntityTargetEvent} when you choose target
```
Requiring a plugin to throw the event feels a bit ugly to me.  Could this instead by handled by the implementation?

natemort commented on src/main/java/org/bukkit/entity/Intelligent.java:
```diff
@@ -0,0 +1,95 @@
+     * 
+     * @param entity
+     * @param speedPitch
+     * @param speedYaw
```
What do these parameters do?

natemort commented on src/main/java/org/bukkit/entity/Intelligent.java:
```diff
@@ -0,0 +1,95 @@
+     * @param y
+     * @param z
+     * @param speedPitch
+     * @param speedYaw
```
What do these parameters do?

natemort commented on src/main/java/org/bukkit/entity/Intelligent.java:
```diff
@@ -0,0 +1,95 @@
+     * @param x
+     * @param y
+     * @param z
+     * @param speed
```
What do these parameters do?

tomsik68 commented:
@evilmidget38 What improvement? Formatting?
@andrepl What else do I need to improve?

natemort commented on src/main/java/org/bukkit/entity/Intelligent.java:
```diff
@@ -0,0 +1,95 @@
+     */
+    public void moveTo(double x,double y,double z,float speed);
+    /** 
+     * Makes entity jump
```
How high is a jump?  Is the velocity applied to the jump?  This needs a better explanation.

natemort commented on src/main/java/org/bukkit/ai/Pathfinder.java:
```diff
@@ -0,0 +1,54 @@
+import org.bukkit.entity.LivingEntity;
+
+/** 
+ * Represents a Pathfinder for {@link Controllable}s
```
Where's this Controllable interface?

andrepl replied:
is this supposed to be 'Intelligent' ?

natemort commented on src/main/java/org/bukkit/ai/Pathfinder.java:
```diff
@@ -0,0 +1,54 @@
+     */
+    public boolean canContinue();
+    /** 
+     * Called when pathfinder execution stops
```
When does the pathfinder execution stop?  What determines this?

natemort commented on src/main/java/org/bukkit/ai/Pathfinder.java:
```diff
@@ -0,0 +1,54 @@
+     */
+    public void onStop();
+    /** 
+     * Gets type of this Pathfinder. Can't be null.
```
Perhaps a brief explanation regarding which type should be returned would be appropriate here.

natemort commented on src/main/java/org/bukkit/ai/PathfinderType.java:
```diff
@@ -0,0 +1,18 @@
+    /**
+     * Goal pathfinders are peaceful and help mobs to swim, panic and lots of more useful things
+     */
+    Goal,
```
enums should be all uppercase.

natemort commented on src/main/java/org/bukkit/entity/Intelligent.java:
```diff
@@ -0,0 +1,95 @@
+ */
+public interface Intelligent {
+    /**
+     * 
```
Methods need a method description as well as a return/parameters description.

andrepl commented:
@tomsik68 there are still a lot of whitespace issues, try to work through the ones that evilmidget38 and I pointed out and then we'll find some more :)  I went through the same thing with my first PR not very long ago, its a bit discouraging, but it's all worth it if the code gets accepted! :)

in general, you need to put blank lines between every method, and remove blank lines at the end of a class after the last method.  your if statements are all cramped like this:

```
if(this=that){
    dothis()
}else{
```

these should be:

```
if (this == that) {
    dothis()
} else {
```

SXRWahrheit commented:
Evilmidget has left you a lot of helpful comments. :)

natemort commented on src/main/java/org/bukkit/entity/Intelligent.java:
```diff
@@ -0,0 +1,95 @@
+     */
+    public void setAvoidSun(boolean noSun);
+    /**
+     * Sets speed of this entity while moving. 1.0 is really fast
```
How fast is really fast?

natemort commented on src/main/java/org/bukkit/entity/LivingEntity.java:
```diff
@@ -13,7 +13,7 @@
  * Represents a living entity, such as a monster or player
  */
-public interface LivingEntity extends Entity, Damageable {
+public interface LivingEntity extends Entity, Damageable, Intelligent {
```
Is this API applicable to truly all LivingEntities?  Is Intelligent the best name for the interface?  I feel like Controllable would make more sense, and that this interface shouldn't be implemented by all Entities(namely Players).

tomsik68 replied:
Ok, I'll rename it back to Controllable. 
I based it on vanilla minecraft, as even players have those controllers there. I'm working on a change...

andrepl commented:
Does this have any effect on mobs that don't use the new AI?  if not, it might be best to add the interface only to those entities that do.

tomsik68 commented:
@andrepl No, it doesn't. I'll do what you say and add it just to entities that can use pathfinders. Thanks for pointing that out ;)

tomsik68 committed:
Pathfinders API was nearly whole rewritten, since not all LivingEntities
support pathfinders. Most of LivingEntities ignore them. Now it's fixed
and PathfinderTest was also updated.

natemort commented on src/main/java/org/bukkit/entity/Skeleton.java:
```diff
@@ -3,7 +3,7 @@
  * Represents a Skeleton.
  */
-public interface Skeleton extends Monster {
+public interface Skeleton extends Monster, PathfindingEntity {
```
Perhaps Monster should extend PathfindingEntity if all of its subinterfaces do.

tomsik68 replied:
Well, they don't... Blaze, EnderDragon, Enderman and lots of others don't...

riking commented:
This seems to me like a pretty thin wrapper around the NMS methods.

Also, you need some kind of way to restore the default behavior once the plugin is done controlling the entity.

riking commented on src/main/java/org/bukkit/ai/PathfinderType.java:
```diff
@@ -0,0 +1,18 @@
+ */
+public enum PathfinderType {
+    /**
+     * Goal pathfinders are peaceful and help mobs to swim, panic and lots of more useful things
```
> Goal pathfinders are not directed at a specific entity. They may be targeted towards a certain block, but not a specific entity. Examples of goal pathfinders would be idling, swimming, panicking, a random stroll, looking at a player, and a sheep eating grass. Note that vanilla treats breeding as a goal pathfinder and not a target pathfinder.

riking commented on src/main/java/org/bukkit/ai/PathfinderType.java:
```diff
@@ -0,0 +1,18 @@
+     */
+    GOAL,
+    /**
+     * Targetting pathfinders help monsters & golems to choose targets
```
> Target pathfinders are directed at a specific entity. Examples of target pathfinders would be monsters attacking a player, or a wolf attacking sheep.
> When you choose a target, you MUST call a {@link EntityTargetEvent}.

riking commented on src/main/java/org/bukkit/entity/PathfindingEntity.java:
```diff
@@ -0,0 +1,18 @@
+     */
+    public void attachPathfinder(Pathfinder pf);
+    /**
+     * Deletes all the pathfinders(even the vanilla ones)
```
> Clear all pathfinders from this entity. You are required to immediately attach additional pathfinders.

#### Create a new method to restore the vanilla pathfinders, please.

riking commented on src/main/java/org/bukkit/entity/Controllable.java:
```diff
@@ -0,0 +1,100 @@
+     * @param speedYaw - Speed of rotation around the Y-axis
+     */
+    public void lookAt(Location location, float speedPitch, float speedYaw);
+    /** 
```
You need an extra linebreak between the method and the next javadoc.

Don't think that's documented anywhere, but do it.

riking commented:
Please create a new method to restore the vanilla pathfinders.

Also, I think that you need to rethink the way you're implementing this. Write it out on paper, and stuff.

tomsik68 commented:
@riking Ok, it seems like you're right as it's too thin for bukkit. I'll rethink a way of implementing and probably submit a new PR as this is already too much commits. Thanks everyone for helping out, I'm now writing new code...
